core/src/main/scala/org/bitcoins/core/psbt/PSBT.scala

```diff
@@ -1407,23 +1469,130 @@ case class InputPSBTMap(elements: Vector[InputPSBTRecord]) extends PSBTMap {
+      }
+    }
+  }
+
+  def fromUTXOSpendingInfo(
+      spendingInfo: UTXOSpendingInfo,
+      unsignedTx: Transaction,
+      nonWitnessTxOpt: Option[Transaction])(
+      implicit ec: ExecutionContext): Future[InputPSBTMap] = {
+    val sigPs = spendingInfo.signers
+      .map(signer => (signer.publicKey, Promise[ECDigitalSignature]()))
+      .toMap
+    val sigFs = {
+      val futures = sigPs.map { case (key, sigP) => (key, sigP.future) }.values
+      val future = Future.sequence(futures)
+
+      future.map(_.zip(sigPs.keys).toMap)
+    }
+
+    val sideEffectSigners = spendingInfo.signers.map { signer =>
+      val signFunction = { bytes: ByteVector =>
+        signer.signFunction(bytes).map { sig: ECDigitalSignature =>
+          val sigWithHashType = ECDigitalSignature(
+            sig.bytes ++ ByteVector.fromByte(spendingInfo.hashType.byte))
+          sigPs(signer.publicKey).success(sigWithHashType)
+
+          sig
+        }
+      }
+
+      Sign(signFunction, signer.publicKey)
+    }
+    val sideEffectSpendingInfo = BitcoinUTXOSpendingInfo(
+      outPoint = spendingInfo.outPoint,
+      output = spendingInfo.output,
+      signers = sideEffectSigners,
+      redeemScriptOpt = spendingInfo.redeemScriptOpt,
+      scriptWitnessOpt = spendingInfo.scriptWitnessOpt,
+      hashType = spendingInfo.hashType,
+      conditionalPath = spendingInfo.conditionalPath
+    )
+
+    val _ = BitcoinSigner.sign(sideEffectSpendingInfo,
+                               unsignedTx,
+                               isDummySignature = false)
+
+    sigFs.map { sigs =>
+      val builder = Vector.newBuilder[InputPSBTRecord]
+
+      spendingInfo match {
+        case _: SegwitV0NativeUTXOSpendingInfo |
+            _: P2SHNestedSegwitV0UTXOSpendingInfo =>
+          builder.+=(WitnessUTXO(spendingInfo.output))
+        case _: RawScriptUTXOSpendingInfo | _: P2SHNoNestSpendingInfo |
+            _: UnassignedSegwitNativeUTXOSpendingInfo =>
+          nonWitnessTxOpt match {
+            case None => ()
+            case Some(nonWitnessTx) =>
+              builder.+=(NonWitnessOrUnknownUTXO(nonWitnessTx))
+          }
+      }
+
+      val sigsAndPubKeys = sigs.map {
+        case (sig, key) => PartialSignature(key, sig)
       }
+      builder.++=(sigsAndPubKeys)
+
+      val sigHashType = SigHashType(spendingInfo.hashType)
+      builder.+=(sigHashType)
+
+      spendingInfo match {
+        case p2sh: P2SHNoNestSpendingInfo =>
+          builder.+=(RedeemScript(p2sh.redeemScript))
+        case p2sh: P2SHNestedSegwitV0UTXOSpendingInfo =>
+          builder.+=(RedeemScript(p2sh.redeemScript))
+          p2sh.scriptWitness match {
+            case p2wsh: P2WSHWitnessV0 =>
+              builder.+=(WitnessScript(p2wsh.redeemScript))
+            case _: P2WPKHWitnessV0 => ()
+          }
+        case p2wsh: P2WSHV0SpendingInfo =>
+          builder.+=(WitnessScript(p2wsh.scriptWitness.redeemScript))
+        case _: BitcoinUTXOSpendingInfo => ()
```

Comment:
can we actually enumerate the different types here so we get warnings when new `UTXOSpendingInfos` are added to the ADT?